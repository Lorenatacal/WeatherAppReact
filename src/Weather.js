import React, { Component } from 'react';
import axios from 'axios';
import kelvinToCelsius from 'kelvin-to-celsius';
import kelvinToFahrenheit from 'kelvin-to-fahrenheit';
import styled from 'styled-components';
import Icon from './Icon';

const WeatherContainer = styled.div`
    border-radius: 20px;
    padding: 0.25em;
    margin: 2em;
    margin-left: 25px;
    margin-right: 25px;
    border: 1px solid lightgrey;
    align: center;
    text-align: center;
    font-family: arial;
`;

const StyledButton = styled.button`
    border-radius: 5px;
    padding: 0.25em;
    margin: 1em;
    border: 0.5px, #007bff;
    background: #007bff;
    color: white;
    text-align: center;
    font-size: 1.1rem;
    font-family: 'Merriweather', serif;
`

const StyledLocation = styled.h2`
    font-family: 'Source Serif Pro', serif;
`

const StyledText = styled.p`
    font-family: 'Merriweather', serif;
    font-size:  1.1rem;
`


class Weather extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            response: false,
            name: '',
            country: '',
            description: '',
            clouds: 0,
            icon: '',
            tempCelsiuse: 0,
            tempCelsiuseI: 0,
            tempFahrenheit: 0,
            minTempCelsiuse: 0,
            minTempFahrenheit: 0,
            maxTempCelsiuse: 0,
            maxTempFahrenheit: 0,
            wind: 0,
            humidity:0,
            showCelsiuse: true,
        };
    }

    handleClickFahrenheit = () => {
        this.setState({
            showCelsiuse: false,
        });
    }

    handleClickCelsiuse = () => {
        this.setState({
            showCelsiuse: true,
        });
    }

    render() {
        if (this.props.coords && !this.state.response){
            axios.get('https://api.openweathermap.org/data/2.5/weather?', {
                params: {
                    lat: this.props.coords.latitude,
                    lon: this.props.coords.longitude,
                    appid: process.env.REACT_APP_WEATHER_API_KEY,
                }
            })
            .then((response) => {
                this.setState({
                    response: true,
                    name: response.data.name,
                    country: response.data.sys.country,
                    description: response.data.weather[0].description,
                    tempCelsiuse: kelvinToCelsius(response.data.main.temp),
                    tempFahrenheit: kelvinToFahrenheit(response.data.main.temp),
                    minTempCelsiuse: kelvinToCelsius(response.data.main.temp_min),
                    minTempFahrenheit: kelvinToFahrenheit(response.data.main.temp_min),
                    maxTempCelsiuse: kelvinToCelsius(response.data.main.temp_max),
                    maxTempFahrenheit: kelvinToFahrenheit(response.data.main.temp_max),
                    clouds: response.data.clouds.all,
                    wind: response.data.wind.speed,
                    humidity: response.data.main.humidity,
                })
            })
        }

                return(
                    <WeatherContainer>
                        <StyledLocation> Today in: {this.state.name}, {this.state.country}</StyledLocation>
                        <Icon description={this.state.description}/>
                        <StyledText> We have {this.state.description}, with {this.state.clouds}% clouds</StyledText>
                        {
                            this.state.showCelsiuse ?  
                                <StyledText>The temperature is: {this.state.tempCelsiuse}°C, with minimum of {this.state.minTempCelsiuse}°C and a maximum of {this.state.maxTempCelsiuse}°C</StyledText> : 
                                <StyledText> The temperature is: {this.state.tempFahrenheit}°F, with minimum of {this.state.minTempFahrenheit}°F and a maximum of {this.state.maxTempFahrenheit}°F</StyledText>
                        }
                        <StyledText> The wind is blowing with: {this.state.wind}m/s and we have a humidity of: {this.state.humidity}%</StyledText>
                        <StyledButton onClick={this.handleClickFahrenheit}>Change °F</StyledButton>
                        <StyledButton onClick={this.handleClickCelsiuse}>Change °C</StyledButton>
                    </WeatherContainer>
                )
    }
}


export default Weather;

