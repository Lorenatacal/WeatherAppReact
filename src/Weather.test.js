import React, { Component } from 'react';
import axios from 'axios';
import Weather from './Weather';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import MockAdapter from 'axios-mock-adapter';


Enzyme.configure({ adapter: new Adapter() });

function flushPromises() {
    return new Promise(resolve => setImmediate(resolve));
}

test("Weather should render correctly without data", () => {
    const wrapper = Enzyme.shallow(<Weather />)
    expect(toJson(wrapper)).toMatchSnapshot();
});

test("Weather should render correctly with data", async () => {
    //we re returning a fake response
    const mockedResponse = {
        name: "London",
        sys: {
            country: "UK",
        },
        weather: [{
            description: "sunny",
        }],
        main: {
            temp: 296,
            temp_min: 293.15,
            temp_max: 296,
            humidity: 10,
        },
        clouds: {
            all: 10,
        },
        wind: {
            speed: 4.1,
        },
    };
    const mock = new MockAdapter(axios);
    mock
        .onGet('https://api.openweathermap.org/data/2.5/weather?')
        .reply(200, mockedResponse);
    // we are mocking the props object
    const coordinates = {
        latitude: 51.48827190000001, 
        longitude: -0.059787999999999994
    }; 
    const wrapper = Enzyme.shallow(<Weather coords={coordinates} />)
    await flushPromises();
    expect(toJson(wrapper)).toMatchSnapshot();
});

test("fahrenheitButton should change the temperature from Celsiuse to Fahrenhite", async () => {
    const mockedResponse = {
        name: "London",
        sys: {
            country: "UK",
        },
        weather: [{
            description: "sunny",
        }],
        main: {
            temp: 296.15,
            temp_min: 293.15,
            temp_max: 296.15,
            humidity: 10,
        },
        clouds: {
            all: 10,
        },
        wind: {
            speed: 4.1,
        },
    };
    const mock = new MockAdapter(axios);
    mock
        .onGet('https://api.openweathermap.org/data/2.5/weather?')
        .reply(200, mockedResponse);

    const coordinates = {
        latitude: 51.48827190000001, 
        longitude: -0.059787999999999994
    }; 

    const wrapper = Enzyme.shallow(<Weather coords={coordinates} />)
    await flushPromises();
    const button = wrapper.find('[data-name="fahrenheitButton"]');
    button.simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();  
});

test("celsiuseButton should change the temperature from Fahrenheit to Celsiuse", async () => {
    const mockedResponse = {
        name: "London",
        sys: {
            country: "UK",
        },
        weather: [{
            description: "sunny",
        }],
        main: {
            temp: 296.15,
            temp_min: 293.15,
            temp_max: 296.15,
            humidity: 10,
        },
        clouds: {
            all: 10,
        },
        wind: {
            speed: 4.1,
        },
    };
    const mock = new MockAdapter(axios);
    mock
        .onGet('https://api.openweathermap.org/data/2.5/weather?')
        .reply(200, mockedResponse);

    const coordinates = {
        latitude: 51.48827190000001, 
        longitude: -0.059787999999999994
    }; 

const wrapper = Enzyme.shallow(<Weather coords={coordinates} />);
await flushPromises();
const celsiuseButton = wrapper.find('[data-name="celsiuseButton"]');
celsiuseButton.simulate('click');
expect(toJson(wrapper)).toMatchSnapshot();
  
});