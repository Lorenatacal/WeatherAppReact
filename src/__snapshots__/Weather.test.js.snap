// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather should render correctly with data 1`] = `
<styled.div>
  <styled.h2>
     Today in: 
    , 
  </styled.h2>
  <Icon
    description=""
  />
  <styled.p>
     We have 
    , with 
    0
    % clouds
  </styled.p>
  <styled.p>
    The temperature is: 
    0
    °C, with minimum of 
    0
    °C and a maximum of 
    0
    °C
  </styled.p>
  <styled.p>
     The wind is blowing with: 
    0
    m/s and we have a humidity of: 
    0
    %
  </styled.p>
  <styled.button
    onClick={[Function]}
  >
    Change °F
  </styled.button>
  <styled.button
    onClick={[Function]}
  >
    Change °C
  </styled.button>
</styled.div>
`;

exports[`Weather should render correctly without data 1`] = `
<styled.div>
  <styled.h2>
     Today in: 
    , 
  </styled.h2>
  <Icon
    description=""
  />
  <styled.p>
     We have 
    , with 
    0
    % clouds
  </styled.p>
  <styled.p>
    The temperature is: 
    0
    °C, with minimum of 
    0
    °C and a maximum of 
    0
    °C
  </styled.p>
  <styled.p>
     The wind is blowing with: 
    0
    m/s and we have a humidity of: 
    0
    %
  </styled.p>
  <styled.button
    onClick={[Function]}
  >
    Change °F
  </styled.button>
  <styled.button
    onClick={[Function]}
  >
    Change °C
  </styled.button>
</styled.div>
`;
